pds_doi_service/core/outputs/osti.py

```diff
@@ -137,3 +169,775 @@ def create_osti_doi_record(self, dois, content_type=CONTENT_TYPE_XML):
+        response text.
+
+        The format of i_url is: https://www.osti.gov/iad2test/api/records/
+        and will be appended by fields in query_dict:
+
+            if query_dict = {'id':14108,'status':'Error'}
+                then https://www.osti.gov/iad2test/api/records?id=14108&status=Error
+
+            if query_dict = {'id'=1327397,'status'='Registered'}
+                then https://www.osti.gov/iad2test/api/records?id=1327397&status=Registered
+        """
+        if content_type not in self._content_type_map:
+            raise ValueError('Invalid content type requested, must be one of '
+                             f'{",".join(list(self._content_type_map.keys()))}')
+
+        auth = HTTPBasicAuth(username, password)
+
+        headers = {
+            'Accept': self._content_type_map[content_type],
+            'Content-Type': self._content_type_map[content_type]
+        }
+
+        # OSTI server requires 'rows' field to know how many max rows to fetch at once.
+        initial_payload = {'rows': self.MAX_TOTAL_ROWS_RETRIEVE}
+
+        # If user provided a query_dict, append to our initial payload.
+        if query:
+            initial_payload.update(query)
+            # Do a sanity check and only fetch valid field names.
+            query_dict = self._validate_field_names(initial_payload)
+        else:
+            query_dict = initial_payload
+
+        logger.debug("initial_payload: %s", initial_payload)
+        logger.debug("query_dict: %s", query_dict)
+        logger.debug("i_url: %s", url)
+
+        osti_response = requests.get(
+            url, auth=auth, params=query_dict, headers=headers
+        )
+
+        try:
+            osti_response.raise_for_status()
+        except requests.exceptions.HTTPError as http_err:
+            # Detail text is not always present, which can cause json parsing
+            # issues
+            details = (
+                f'Details: {pprint.pformat(json.loads(osti_response.text))}'
+                if osti_response.text else ''
+            )
+
+            raise WebRequestException(
+                'DOI submission request to OSTI service failed, '
+                f'reason: {str(http_err)}\n{details}'
+            )
+
+        return osti_response.text
+
+    def _validate_field_names(self, query_dict):
+        """
+        Validates the provided fields by the user to make sure they match the
+        expected fields by OSTI:
+
+            https://www.osti.gov/iad2test/docs#endpoints-recordlist
+
+        """
+        o_validated_dict = {}
+
+        for key in query_dict:
+            # If the key is valid, save the field and value to return.
+            if key in self.ACCEPTABLE_FIELD_NAMES_LIST:
+                o_validated_dict[key] = query_dict[key]
+            else:
+                logger.error(f"Unexpected field name '{key}' in query_dict")
+                exit(1)
+
+        return o_validated_dict
+
+
+class DOIOstiXmlWebParser(DOIOstiWebParser):
```

Comment:
I am trying to have one class per file so that it is easier to understand the content of a module. Do you have a different strategy on that I would be interested to know your point of view.

Anyhow this is not something critical to change here.

Reply:
I thought it might be nice to have everything Osti-related confined to a single module, but there ended up being more code in there than I thought, so I'm inclined to break it up into per-class modules as you've suggested. I will tackle that as part of the next PR.

Reply:
Ok, sounds good. Thanks @collinss-jpl